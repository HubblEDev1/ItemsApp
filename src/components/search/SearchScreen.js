import React, { useMemo } from 'react'
import { useLocation } from 'react-router';
import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm/useForm';
import { HeroCard } from '../heroes/HeroCard';
import queryString from 'query-string';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();
    //console.log(location.search);
    const { q = '' } = queryString.parse(location.search);

    const [{hero}, handleInputChange, reset] = useForm({
        hero: q
    });


    const heroesFiltered = useMemo(() => getHeroesByName( q ), [q]);  

    const handleSearch = (e) => {
        e.preventDefault();
        if( hero.trim().length <= 1){
            return;
        }

        history.push(`?q=${hero}`)
    
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <hr/>
                    <form onSubmit={handleSearch}> 
                        <input
                            type="text"
                            name="hero"
                            placeholder="Find your hero"
                            className="form-control"
                            onChange={handleInputChange}
                            value={hero}
                        />
                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4> Results </h4>
                    <hr/>
                    
                    {
                        (q === '')
                            && 
                            <div className="alert alert-info">
                                Search a hero
                            </div>
                    }

                    {
                        (q !== '' && heroesFiltered.length === 0)
                            && 
                            <div className="alert alert-danger">
                                There isn't a hero with {q}
                            </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
                
            </div>
        </div>
    )
}
