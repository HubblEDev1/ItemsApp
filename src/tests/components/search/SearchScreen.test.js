import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe('Testing <SearchScreen/>', () => {
    test('should show the snapshot with correct values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path = "/search" component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');

    })

    test('should show to batman and input with queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path = "/search" component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })

    test('should show the error if doesnt find the hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123123123']}>
                <Route path = "/search" component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper.find('.alert-danger').text().trim()).toBe('There isn\'t a hero with batman123123123');
        expect(wrapper).toMatchSnapshot();
    })

    test('should call the push', () => {
        const history = {
            push: jest.fn()
        }
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123123123']}>
                <Route 
                    path = "/search" 
                    component={() => <SearchScreen history={history}/>} />
            </MemoryRouter>
        )
        
        wrapper.find('input').simulate('change', {
            target: {
                name: 'hero', 
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })

        expect(history.push).toHaveBeenCalledWith('?q=batman');
    })
    
    
})
