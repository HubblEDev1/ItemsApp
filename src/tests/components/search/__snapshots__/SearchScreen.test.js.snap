// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <SearchScreen/> should show the error if doesnt find the hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123123123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There isn't a hero with 
        batman123123123
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing <SearchScreen/> should show the snapshot with correct values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing <SearchScreen/> should show to batman and input with queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img animate__animated animate__fadeInLeft"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
